README.md
# Flutter progress indicators

With this project, we are aiming to provide some cool, animated progress indicators to Flutter developers. 

# Currently available
## Sample
Here's how it looks like.

![](imgs/sample.gif)

## Usage
Import `package:progress_indicators/progress_indicators.dart` and use these avaiable widgets:
* `FadingText` 
* `JumpingText`
* `ScalingText`
* `JumpingDots`
* `HeartbeatProgressIndicator`
* `GlowingProgressIndicator`
* `CollectionSlideTranstion`
* `CollectionScaleTransition` 

## Example
```
Widget build(BuildContext context) {
    print(Offset(0.0, -1.0).distanceSquared - Offset(0.0, 0.0).distanceSquared);
    return new Scaffold(
      appBar: new AppBar(
        title: new Text(widget.title),
      ),
      body: new Center(
        child: new Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            new Text('Jumping dots'),
            JumpingDotsProgressIndicator(
              fontSize: 20.0,
            ),
            SizedBox(height: 60.0),
            new Text('Heartbeat'),
            SizedBox(height: 16.0),
            HeartbeatProgressIndicator(
              child: Icon(Icons.home),
            ),
            SizedBox(height: 60.0),
            new Text('Glowing'),
            GlowingProgressIndicator(
              child: Icon(Icons.home),
            ),
            SizedBox(height: 32.0),
            FadingText('Loading...'),
            SizedBox(height: 32.0),
            JumpingText('Loading...'),
            SizedBox(height: 32.0),
            ScalingText('Loading...'),
            SizedBox(height: 32.0),
            CollectionSlideTransition(
              children: <Widget>[
                Icon(Icons.android),
                Icon(Icons.apps),
                Icon(Icons.announcement),
              ],
            ),
            SizedBox(height: 32.0),
            CollectionScaleTransition(
              children: <Widget>[
                Icon(Icons.android),
                Icon(Icons.apps),
                Icon(Icons.announcement),
              ],
            ),
          ],
        ),
      ),
    );
  }
```

For complete sample, see the [this example app](https://github.com/wal33d006/progress_indicators/tree/master/example).

CHANGELOG.md
## 1.0.0

* Support for null-safety

## 0.1.5

* Fix not animation when font size is increased.
* Update example with Android X capabilities.

## 0.1.4

* Allows you to specify a custom text style for your text effects.

## 0.1.3

* Fixes JumpingText to actually make use of the "end" parameter.

## 0.1.1

* Fixes `pubspec.yaml` to remove Dart 2 support issue.

## 0.1.0

* First release that contains `FadingText`, `JumpingText`, `ScalingText`, `JumpingDots`, `HeartbeatProgressIndicator`, `GlowingProgressIndicator`, `CollectionSlideTranstion` and `CollectionScaleTransition` widgets. 


example/test/widget_test.dart
// This is a basic Flutter widget test.
// To perform an interaction with a widget in your test, use the WidgetTester utility that Flutter
// provides. For example, you can send tap and scroll gestures. You can also use WidgetTester to
// find child widgets in the widget tree, read text, and verify that the values of widget properties
// are correct.

import '../lib/main.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(new MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


test/progress_indicators_test.dart
// import 'package:test/test.dart';

// void main() {
//   test('adds one to input values', () {
    
//   });
// }


lib/src/jumping_dots.dart
import 'package:flutter/material.dart';
import 'package:flutter/animation.dart';

/// Adds a horizontal list of variable number of jumping dots
///
/// The animation is a smooth up/down continuous animation of each dot.
/// This animation can be used where a text is being expected from an async call
/// The below class is a private [AnimatedWidget] class which is called in the
/// [StatefulWidget].
class _JumpingDot extends AnimatedWidget {
  final Color? color;
  final double? fontSize;

  _JumpingDot({Key? key, required Animation<double> animation, this.color, this.fontSize})
      : super(key: key, listenable: animation);

  Widget build(BuildContext context) {
    final Animation<double> animation = listenable as Animation<double>;
    return Container(
      height: animation.value + fontSize!,
      child: Text(
        '.',
        style: TextStyle(color: color, fontSize: fontSize),
      ),
    );
  }
}

/// Creates a list with [numberOfDots] text dots, with 3 dots as default
/// default [fontSize] of 10.0, default [color] as black, [dotSpacing] (gap
/// between each dot) as 0.0 and default time for one cycle of animation
/// [milliseconds] as 250.
/// One cycle of animation is one complete round of a dot animating up and back
/// to its original position.
class JumpingDotsProgressIndicator extends StatefulWidget {
  /// Number of dots that are added in a horizontal list, default = 3.
  final int numberOfDots;

  /// Font size of each dot, default = 10.0.
  final double fontSize;

  /// Spacing between each dot, default 0.0.
  final double dotSpacing;

  /// Color of the dots, default black.
  final Color color;

  /// Time of one complete cycle of animation, default 250 milliseconds.
  final int milliseconds;

  /// Starting and ending values for animations.
  final double beginTweenValue = 0.0;
  double endTweenValue = 8.0;

  /// Creates a jumping do progress indicator.
  JumpingDotsProgressIndicator({
    this.numberOfDots = 3,
    this.fontSize = 10.0,
    this.color = Colors.black,
    this.dotSpacing = 0.0,
    this.milliseconds = 250,
  });

  _JumpingDotsProgressIndicatorState createState() =>
      _JumpingDotsProgressIndicatorState(
        numberOfDots: this.numberOfDots,
        fontSize: this.fontSize,
        color: this.color,
        dotSpacing: this.dotSpacing,
        milliseconds: this.milliseconds,
      );
}

class _JumpingDotsProgressIndicatorState
    extends State<JumpingDotsProgressIndicator> with TickerProviderStateMixin {
  int? numberOfDots;
  int? milliseconds;
  double? fontSize;
  double? dotSpacing;
  Color? color;
  List<AnimationController> controllers = <AnimationController>[];
  List<Animation<double>> animations = <Animation<double>>[];
  List<Widget> _widgets = <Widget>[];

  _JumpingDotsProgressIndicatorState({
    this.numberOfDots,
    this.fontSize,
    this.color,
    this.dotSpacing,
    this.milliseconds,
  });

  initState() {
    super.initState();
    for (int i = 0; i < numberOfDots!; i++) {
      _addAnimationControllers();
      _buildAnimations(i);
      _addListOfDots(i);
    }

    controllers[0].forward();
  }

  void _addAnimationControllers() {
    controllers.add(AnimationController(
        duration: Duration(milliseconds: milliseconds!), vsync: this));
  }

  void _addListOfDots(int index) {
    _widgets.add(
      Padding(
        padding: EdgeInsets.only(right: dotSpacing!),
        child: _JumpingDot(
          animation: animations[index],
          fontSize: fontSize,
          color: color,
        ),
      ),
    );
  }

  void _buildAnimations(int index) {
    animations.add(
      Tween(begin: widget.beginTweenValue, end: widget.endTweenValue)
          .animate(controllers[index])
            ..addStatusListener(
              (AnimationStatus status) {
                if (status == AnimationStatus.completed)
                  controllers[index].reverse();
                if (index == numberOfDots! - 1 &&
                    status == AnimationStatus.dismissed) {
                  controllers[0].forward();
                }
                if (animations[index].value > widget.endTweenValue / 2 &&
                    index < numberOfDots! - 1) {
                  controllers[index + 1].forward();
                }
              },
            ),
    );
  }

  Widget build(BuildContext context) {
    return SizedBox(
      height: fontSize! + (fontSize! * 0.5),
      child: Row(
        mainAxisAlignment: MainAxisAlignment.center,
        children: _widgets,
      ),
    );
  }

  dispose() {
    for (int i = 0; i < numberOfDots!; i++) controllers[i].dispose();
    super.dispose();
  }
}


lib/src/animated_text.dart
import 'package:flutter/material.dart';

import 'package:progress_indicators/src/collection_animators.dart';

/// Adds fading effect on each character in the [text] provided to it.
///
/// The animation is repeated continuously so this widget is ideal
/// to be used as progress indicator.
/// Although this widget does not put explicit limit on string character count,
/// however, it should be given such that it does not exceed a line.
///
/// The text displayed follows the default [TextStyle] of current theme, unless
/// otherwise specified.
class FadingText extends StatefulWidget {
  /// Text to animate
  final String text;
  /// Custom text style. If not specified, uses the default style.
  final TextStyle? style;

  /// Creates a fading continuous animation.
  ///
  /// The provided [text] is continuously animated using [FadeTransition].
  /// [text] must not be null.
  FadingText(this.text, {this.style}) : assert(text != null);

  @override
  _FadingTextState createState() => new _FadingTextState();
}

class _FadingTextState extends State<FadingText> with TickerProviderStateMixin {
  final _characters = <MapEntry<String, Animation>>[];
  late AnimationController _controller;

  @override
  void initState() {
    super.initState();

    _controller = AnimationController(
      vsync: this,
      duration: Duration(seconds: 2),
    );

    var start = 0.2;
    final duration = 0.6 / widget.text.length;
    widget.text.runes.forEach((int rune) {
      final character = new String.fromCharCode(rune);
      final animation = Tween(
        begin: 0.0,
        end: 1.0,
      ).animate(
        CurvedAnimation(
          curve: Interval(start, start + duration, curve: Curves.easeInOut),
          parent: _controller,
        ),
      );
      _characters.add(MapEntry(character, animation));
      start += duration;
    });

    _controller.addStatusListener((status) {
      if (status == AnimationStatus.completed) {
        _controller.reverse();
      } else if (status == AnimationStatus.dismissed) {
        _controller.forward();
      }
    });
    _controller.forward();
  }

  @override
  Widget build(BuildContext context) {
    return Row(
      mainAxisSize: MainAxisSize.min,
      children: _characters
          .map(
            (entry) => FadeTransition(
                  opacity: entry.value as Animation<double>,
                  child: Text(entry.key, style: widget.style),
                ),
          )
          .toList(),
    );
  }

  dispose() {
    _controller.dispose();
    super.dispose();
  }
}

/// Adds jumping effect on each character in the [text] provided to it.
///
/// The animation is repeated continuously so this widget is ideal
/// to be used as progress indicator.
/// Although this widget does not put explicit limit on string character count,
/// however, it should be given such that it does not exceed a line.
///
/// The text displayed follows the default [TextStyle] of current theme, unless
/// otherwise specified.
class JumpingText extends StatelessWidget {
  final String text;
  final Offset begin = Offset(0.0, 0.0);
  final Offset end;
  final TextStyle? style;

  /// Creates a jumping text widget.
  ///
  /// Each character in [text] is animated to look like a jumping effect.
  /// The [end] is the target [Offset] for each character.
  JumpingText(this.text, {this.end = const Offset(0.0, -0.5), this.style});

  @override
  Widget build(BuildContext context) {
    return CollectionSlideTransition(
      end: end,
      children: text.runes
          .map(
            (rune) => Text(String.fromCharCode(rune), style: style),
          )
          .toList(),
    );
  }
}

/// Adds jumping effect on each character in the [text] provided to it.
///
/// The animation is repeated continuously so this widget is ideal
/// to be used as progress indicator.
/// Although this widget does not put explicit limit on string character count,
/// however, it should be given such that it does not exceed a line.
///
/// The text displayed follows the default [TextStyle] of current theme, unless
/// otherwise specified.
class ScalingText extends StatelessWidget {
  /// The text to add scaling effect to.
  final String text;
  final double begin = 1.0;
  final double end;
  final TextStyle? style;

  /// Creates a jumping text widget.
  ///
  /// Each character in [text] is scaled to [end].
  ScalingText(this.text, {this.end = 2.0, this.style});

  @override
  Widget build(BuildContext context) {
    return CollectionScaleTransition(
      end: end,
      children: text.runes
          .map(
            (rune) => Text(String.fromCharCode(rune), style: style),
          )
          .toList(),
    );
  }
}


lib/src/simple_indicators.dart
import 'package:flutter/material.dart';

/// Adds a heartbeat animation to its child widget.
///
/// The animation is repeated continuously.
class HeartbeatProgressIndicator extends StatefulWidget {
  /// The duration of animation playback for each iteration.
  final Duration duration;

  /// Starting scale of child widget. Normally, it should be 1.0.
  final double startScale;

  /// Ending scale of child widget.
  final double endScale;

  /// Widget to apply the animation to.
  final Widget child;

  /// Creates heartbeat indicator.
  ///
  /// [child] widget must not be null.
  /// [startScale] is optional and default value is 1.0.
  /// [endScale] is optional and default value is 2.0.
  /// [duration] is optional and default value is 1 second.
  HeartbeatProgressIndicator({
    this.duration = const Duration(seconds: 1),
    this.startScale: 1.0,
    this.endScale: 2.0,
    required this.child,
  }) : assert(child != null);

  @override
  _HeartbeatProgressIndicatorState createState() =>
      _HeartbeatProgressIndicatorState();
}

class _HeartbeatProgressIndicatorState extends State<HeartbeatProgressIndicator>
    with TickerProviderStateMixin {
  late Animation _animation;
  late AnimationController _controller;

  initState() {
    super.initState();
    _controller = _createOscillatingAnimationController(widget.duration, this);
    _animation = _createTweenAnimation(
      _controller,
      begin: widget.startScale,
      end: widget.endScale,
      curve: Curves.elasticOut,
    );
    _controller.forward();
  }

  @override
  Widget build(BuildContext context) =>
      ScaleTransition(scale: _animation as Animation<double>, child: widget.child);

  dispose() {
    _controller.dispose();
    super.dispose();
  }
}

/// Adds a glowing animation to its child widget.
///
/// The animation is repeated continuously.
class GlowingProgressIndicator extends StatefulWidget {
  /// The duration of animation playback for each iteration.
  final Duration duration;

  /// Widget to apply the animation to.
  final Widget child;

  /// Creates glowing indicator.
  ///
  /// [child] widget must not be null.
  /// [duration] is optional and default value is 1 second.
  GlowingProgressIndicator({
    this.duration = const Duration(seconds: 1),
    required this.child,
  }) : assert(child != null);

  @override
  _GlowingProgressIndicatorState createState() =>
      _GlowingProgressIndicatorState();
}

class _GlowingProgressIndicatorState extends State<GlowingProgressIndicator>
    with TickerProviderStateMixin {
  late Animation _animation;
  late AnimationController _controller;

  initState() {
    super.initState();
    _controller = _createOscillatingAnimationController(widget.duration, this);
    _animation = _createTweenAnimation(
      _controller,
      begin: 0.0,
      end: 1.0,
      curve: Curves.easeOut,
    );
    _controller.forward();
  }

  @override
  Widget build(BuildContext context) =>
      FadeTransition(opacity: _animation as Animation<double>, child: widget.child);

  dispose() {
    _controller.dispose();
    super.dispose();
  }
}

AnimationController _createOscillatingAnimationController(
  Duration duration,
  TickerProvider provider,
) {
  final AnimationController controller =
      AnimationController(duration: duration, vsync: provider);
  controller.addStatusListener((status) {
    if (status == AnimationStatus.completed) {
      controller.reverse();
    } else if (status == AnimationStatus.dismissed) {
      controller.forward();
    }
  });
  return controller;
}

Animation _createTweenAnimation(
  AnimationController controller, {
  double begin: 1.0,
  double end: 2.0,
  Curve curve: Curves.easeOut,
}) {
  return Tween(begin: begin, end: end)
      .animate(CurvedAnimation(parent: controller, curve: curve));
}


lib/src/collection_animators.dart
import 'package:flutter/material.dart';

/// Adds a reversable and repetitive slide transition to each child.
///
/// This widget arranges its [children] in a row and applies forward and
/// reverse animations to each child.
///
/// The animation can be played once or forever and is controlled
/// through [repeat] property.
class CollectionSlideTransition extends StatefulWidget {
  /// Collection of widgets on which slide animation is applied.
  ///
  /// Preferably, [Text], [Icon] or [Image] should be used.
  final List<Widget> children;

  /// End displacement for each child.
  final Offset end;

  /// Start of displacement.
  ///
  /// This is immutable field.
  final Offset begin = Offset.zero;

  /// The toggle to make the animation repeating or non-repeating.
  final bool repeat;

  /// Creates transiton widget.
  ///
  /// [children] is requied and must not be null.
  /// [end] property has default displacement of -1.0 in vertical direction.
  CollectionSlideTransition({
    required this.children,
    this.end = const Offset(0.0, -1.0),
    this.repeat = true,
  }) : assert(children != null);

  @override
  _CollectionSlideTransitionState createState() =>
      new _CollectionSlideTransitionState();
}

class _CollectionSlideTransitionState extends State<CollectionSlideTransition>
    with SingleTickerProviderStateMixin {
  AnimationController? _controller;
  List<_WidgetAnimations<Offset>> _widgets = [];

  @override
  void initState() {
    super.initState();

    _controller = AnimationController(
      vsync: this,
      duration: Duration(seconds: (widget.children.length * 0.25).round()),
    );

    _widgets = _WidgetAnimations.createList<Offset>(
      widgets: widget.children,
      controller: _controller,
      forwardCurve: Curves.ease,
      reverseCurve: Curves.ease,
      begin: widget.begin,
      end: widget.end,
    );

    if (widget.repeat) {
      _controller!.repeat();
    } else {
      _controller!.forward();
    }
  }

  @override
  Widget build(BuildContext context) {
    final end = widget.end;
    return Row(
      mainAxisSize: MainAxisSize.min,
      children: _widgets.map(
        (widgetAnimation) {
          return AnimatedBuilder(
            animation: _controller!,
            builder: (context, _) {
              return FractionalTranslation(
                translation: widgetAnimation.forward!.value.distanceSquared >=
                        end.distanceSquared
                    ? widgetAnimation.reverse!.value
                    : widgetAnimation.forward!.value,
                child: widgetAnimation.widget,
              );
            },
          );
        },
      ).toList(),
    );
  }

  @override
  void dispose() {
    _controller!.dispose();
    super.dispose();
  }
}

/// Adds a reversable and repetitive scale transition to each child.
///
/// This widget arranges its [children] in a row and applies forward and
/// reverse animations to each child.
///
/// The animation can be played once or forever and is controlled
/// through [repeat] property.
class CollectionScaleTransition extends StatefulWidget {
  /// Collection of widgets on which slide animation is applied.
  ///
  /// Preferably, [Text], [Icon] or [Image] should be used.
  final List<Widget> children;

  /// End scale of each child.
  final double end;

  /// Start scale of each child.
  final double begin = 1.0;

  /// The toggle to make the animation repeating or non-repeating.
  final bool repeat;

  /// Creates transiton widget.
  ///
  /// [children] is requied and must not be null.
  /// [end] property has default value of 2.0.
  CollectionScaleTransition({
    required this.children,
    this.end = 2.0,
    this.repeat = true,
  }) : assert(children != null);

  @override
  _CollectionScaleTransitionState createState() =>
      new _CollectionScaleTransitionState();
}

class _CollectionScaleTransitionState extends State<CollectionScaleTransition>
    with SingleTickerProviderStateMixin {
  AnimationController? _controller;
  List<_WidgetAnimations<double>> _widgets = [];

  @override
  void initState() {
    super.initState();

    _controller = AnimationController(
      vsync: this,
      duration: Duration(seconds: (widget.children.length * 0.25).round()),
    );

    _widgets = _WidgetAnimations.createList<double>(
      widgets: widget.children,
      controller: _controller,
      forwardCurve: Curves.easeOut,
      reverseCurve: Curves.easeIn,
      begin: widget.begin,
      end: widget.end,
    );

    if (widget.repeat) {
      _controller!.repeat();
    } else {
      _controller!.forward();
    }
  }

  @override
  Widget build(BuildContext context) {
    final end = widget.end;
    return Row(
      mainAxisSize: MainAxisSize.min,
      children: _widgets.map(
        (widgetAnimation) {
          return AnimatedBuilder(
            animation: _controller!,
            builder: (context, _) {
              return Transform.scale(
                scale: widgetAnimation.forward!.value >= end
                    ? widgetAnimation.reverse!.value
                    : widgetAnimation.forward!.value,
                child: widgetAnimation.widget,
              );
            },
          );
        },
      ).toList(),
    );
  }

  @override
  void dispose() {
    _controller!.dispose();
    super.dispose();
  }
}

class _WidgetAnimations<T> {
  final Widget? widget;
  final Animation<T>? forward;
  final Animation<T>? reverse;

  _WidgetAnimations({this.widget, this.forward, this.reverse});

  static List<_WidgetAnimations<S>> createList<S>({
    required List<Widget> widgets,
    required AnimationController? controller,
    Cubic forwardCurve = Curves.ease,
    Cubic reverseCurve = Curves.ease,
    S? begin,
    S? end,
  }) {
    final animations = <_WidgetAnimations<S>>[];

    var start = 0.0;
    final duration = 1.0 / (widgets.length * 2);
    widgets.forEach((childWidget) {
      final animation = Tween<S>(
        begin: begin,
        end: end,
      ).animate(
        CurvedAnimation(
          curve: Interval(start, start + duration, curve: Curves.ease),
          parent: controller!,
        ),
      );

      final revAnimation = Tween<S>(
        begin: end,
        end: begin,
      ).animate(
        CurvedAnimation(
          curve: Interval(start + duration, start + duration * 2,
              curve: Curves.ease),
          parent: controller,
        ),
      );

      animations.add(_WidgetAnimations(
        widget: childWidget,
        forward: animation,
        reverse: revAnimation,
      ));

      start += duration;
    });

    return animations;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:progress_indicators/progress_indicators.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Progress Indicators',
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(title: 'Progress Indicators'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title}) : super(key: key);

  final String? title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    print(Offset(0.0, -1.0).distanceSquared - Offset(0.0, 0.0).distanceSquared);
    return new Scaffold(
      appBar: new AppBar(
        title: new Text(widget.title!),
      ),
      body: new Center(
        child: new Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            new Text('Jumping dots'),
            JumpingDotsProgressIndicator(
              fontSize: 20.0,
            ),
            SizedBox(height: 60.0),
            new Text('Heartbeat'),
            SizedBox(height: 16.0),
            HeartbeatProgressIndicator(
              child: Icon(Icons.home),
            ),
            SizedBox(height: 60.0),
            new Text('Glowing'),
            GlowingProgressIndicator(
              child: Icon(Icons.home),
            ),
            SizedBox(height: 32.0),
            FadingText('Loading...'),
            SizedBox(height: 32.0),
            JumpingText('Loading...'),
            SizedBox(height: 32.0),
            ScalingText('Loading...'),
            SizedBox(height: 32.0),
            CollectionSlideTransition(
              children: <Widget>[
                Icon(Icons.android),
                Icon(Icons.apps),
                Icon(Icons.announcement),
              ],
            ),
            SizedBox(height: 32.0),
            CollectionScaleTransition(
              children: <Widget>[
                Icon(Icons.android),
                Icon(Icons.apps),
                Icon(Icons.announcement),
              ],
            ),
          ],
        ),
      ),
    );
  }
}